From e12c9269fa9f75d022a3166d8d2a18457c2f8b2d Mon Sep 17 00:00:00 2001
From: Stanislaw Kardach <stanislaw.kardach@cavium.com>
Date: Thu, 18 Oct 2018 10:46:22 +0300
Subject: [PATCH 0317/1051] octeontx: fix double free of octeontx-rm class

There is an invalid destruction (double-free) of struct class
(for octeontx-rm) during octeontx module cleanup. Class created with
class_alloc() has to be destroyed with class_destroy() only.
class_unregister() is only for static class objects.
Underneath class_destroy() calls class_unregister(), hence we have
double-free. After several tries, it results in /sys/class kobject to
be removed and kobject_add_internal fails when trying to find a parent
object for a newly created octeontx-rm class in octeontx module init.

Signed-off-by: Stanislaw Kardach <stanislaw.kardach@cavium.com>
Signed-off-by: Yury Norov <ynorov@marvell.com>
Signed-off-by: Yury Norov <ynorov@caviumnetworks.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
index 35a058280da9..676a35623491 100644
--- a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
+++ b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
@@ -1360,7 +1360,6 @@ static int __init octeontx_init_module(void)
 	device_destroy(octtx_class, octtx_dev);
 
 device_create_err:
-	class_unregister(octtx_class);
 	class_destroy(octtx_class);
 
 class_create_err:
@@ -1412,7 +1411,6 @@ static void __exit octeontx_cleanup_module(void)
 
 	octtx_sysfs_remove(octtx_device);
 	device_destroy(octtx_class, octtx_dev);
-	class_unregister(octtx_class);
 	class_destroy(octtx_class);
 
 	cdev_del(octtx_cdev);
-- 
2.17.1

